### ISSUE TITLE
How to set the imgsz for predict()?

### ISSUE THREAD

**Issue by Avencheng** [ultralytics/ultralytics#16026]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi
I have a trained model ready to use.
```
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8m-cls.pt') 

# Train the model
results = model.train(data="images1", epochs=250, imgsz=640, device=0)

# Predict image
yolo_model = YOLO("runs/classify/train_8m/weights/best.pt")
results = yolo_model.predict('images1/test/News/663.png', imgsz=(352, 640))
```
But the log shows the inference image size is still the same as my training image size
`Speed: 12.1ms preprocess, 13.6ms inference, 0.0ms postprocess per image at shape (1, 3, 640, 640)`

The resolution of my input image for prediction is 1920(width) x 1080(high). 
According to the imgsz parameter format (height, width) for the predict function, i would like to use 352 as my image height and 640 as my image width, in order to maintains a good image aspect ratio and speed up the inference time.

Did I miss any steps? Thanks

**Comment by Y-T-G** [comment]
You can apply it manually.

```
from ultralytics.data.augment import classify_transforms
from PIL import Image

transform = classify_transforms((352, 640))
img = transform(Image.open("ultralytics/assets/bus.jpg"))[None]
result = model.predict(img)
```

**Comment by Y-T-G** [comment]
@ambitious-octopus It seems classification doesn't support overriding the `imgsz` at all. Because it defaults to using the `transforms` stored in the `model`:
https://github.com/ultralytics/ultralytics/blob/88102eb5085a668fff0171fdaa6297af65c2d010/ultralytics/engine/predictor.py#L188-L197

So even if you pass `imgsz=640`, it will still use `224`:
```python
from ultralytics import YOLO
model = YOLO("yolov8n-cls.pt")
result = model("ultralytics/assets/bus.jpg", imgsz=640) # Still uses 224
```

I'm not sure why this was done. Is it for some legacy reasons? Because I see something about legacy transforms in `ClassificationPredictor`:
https://github.com/ultralytics/ultralytics/blob/88102eb5085a668fff0171fdaa6297af65c2d010/ultralytics/models/yolo/classify/predict.py#L34

And `classify_transforms` seems to support non-square size:
https://github.com/ultralytics/ultralytics/blob/88102eb5085a668fff0171fdaa6297af65c2d010/ultralytics/data/augment.py#L2358-L2364

So there's no need to only pass `self.imgsz[0]`. It can use the whole tuple:
https://github.com/ultralytics/ultralytics/blob/88102eb5085a668fff0171fdaa6297af65c2d010/ultralytics/engine/predictor.py#L193

**Comment by Avencheng** [comment]
Hi! Thanks for the reply.
I tried your method, it works! 
Regarding the predict `imgsz`, i think it would be the same as the training `imgsz` you given to it. means not always 224.
```
from ultralytics import YOLO

model = YOLO('yolov8m-cls.pt') 
results = model.train(data="images1", epochs=250, imgsz=640, device=0)
yolo_model = YOLO("runs/classify/train_8m/weights/best.pt")
results = yolo_model.predict('images1/test/News/663.png')
```
As above, I used `imgsz=64` for training and not specify `imgsz` when inferring. the log shows the image shape is (1, 3, 640, 640)
`Speed: 18.8ms preprocess, 16.4ms inference, 0.0ms postprocess per image at shape (1, 3, 640, 640)`

​In addition, Is it appropriate to change the `imgsz` of the predict mode to be different from the training `imgsz`? Will it harm the accuracy? I roughly tested the image size on (384, 640), (416, 640) , (448, 640) and (640, 640), can't really tell which one is better, they perform slightly differently in different categories. Do you have any relevant tests for this ?

Have a nice weekend!

**Comment by Y-T-G** [comment]
It's best to use the same imgsz as training

**Comment by YOOHYOJEONG** [comment]
@Y-T-G 

Hi.
I don't think "imgsz" is working properly in `predict` either.

my predict code
```python
from ultralytics import YOLO

# Load a model
model = YOLO("/ultralytics/DATA/weights/coco_pretrained/yolov11/yolo11n.pt")  # pretrained YOLO11n model

# Run batched inference on a list of images
results = model("/ultralytics/DATA/COCO/val/images", imgsz=(640, 640), conf=0.5)  # return a list of Results objects
```   

I set imgsz to (640, 640), but it's still getting different values in the output.   

![image](https://github.com/user-attachments/assets/6b809401-c0bc-4844-9e37-83543d86c029)
   

Even if you leave it as `imgsz=640`, the output results are the same.

**Comment by Willjay90** [comment]
@Y-T-G 
I'm having the same issue with only `pytorch` model
```
model = YOLO("yolo11s.pt")
model.predict(parser.videos, save=True)

# video 1/1 (frame 10/44) /videos/004.mp4: 384x640 2 persons, 1 chair, 1 remote, 1 cell phone, 4.6ms
```
Also, I tried to convert the model to `torchscript` the image size looks right, but the detection would be slightly different.
```
model = YOLO("yolo11s.pt")
model.export(format="torchscript")

model = YOLO("yolo11s.torchscript")
model.predict(parser.videos, save=True)
# video 1/1 (frame 10/44) /videos/004.mp4: 640x640 2 persons, 1 chair, 1 cell phone, 5.6ms
```

**Comment by Y-T-G** [comment]
@Willjay90 For PyTorch inference or exports with `dynamic`, the images are resized to the minimum rectangle.

https://github.com/ultralytics/ultralytics/issues/16339#issuecomment-2358362884

**Comment by Willjay90** [comment]
@Y-T-G can you elaborate a little bit more.
when we train the model, take image size `(1280, 720)` for example. If we train the model with default config (`imgsz=640`), will the image size during training be `384x640` or `640x640`?

From my understanding, you were mentioning both model resize the image to `360x640` then have different padding.
Since there's slight difference in the result with different padding. What would be the recommended padding strategy? (pad it to 384 or 640?)

**Comment by Y-T-G** [comment]
It trains at 640x640 by default.

The minimum rectangle padding is used for faster inference speed since the image would be smaller as opposed when it's padded fully to 640x640.

**Comment by Willjay90** [comment]
thanks for the clarification 🙂

**Comment by glenn-jocher** [comment]
You're welcome! Let us know if you have any further questions.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#16630) by Y-T-G, with valuable contributions from Laughing-q and glenn-jocher.

Key improvements from this update include:

- Improved image classification transforms to ensure correct preprocessing for non-square image sizes.
- Enhanced dependency handling in the similarity search feature, reducing import-related errors.
- More consistent image size management by leveraging model export metadata when available.
- Updated the Ultralytics package version to 8.3.138.

These enhancements are designed to provide more reliable classification results and a smoother experience with Ultralytics models.

To test if these changes resolve your issue, please upgrade to the latest version using:
```
pip install -U ultralytics>=8.3.138
```
Alternatively, you can install the latest code directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You can also wait for the next official PyPI release if you prefer.

Please let us know if the update addresses your concerns or if you encounter any further issues. Thanks for reporting this, and any additional feedback is always welcome! 🙏
